### PR TITLE
WIP: Implement password hashing

### LINKED ISSUE #36: stop storing plaintext passwords
It's really unacceptable that Axiom stores passwords in plain text.  We should hash them according to best practices (i.e. let's not home-brew some nonsense for obscuring them, let's use a legit PBKDF).

https://github.com/lvh/txscrypt has a pretty great API for this, but it does introduce a Deferred where we didn't have one before.

### PR DESCRIPTION
**Pull request by jsza**
Fixes #36.

Resurrected version of #90; should not be merged before the new APIs have been included in a release.

### CODE REVIEW

File: axiom/test/historic/test_account2to3.py [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+from axiom.test.historic import stubloader
+from axiom.errors import BadCredentials
+from axiom.userbase import getTestContext
+
+SECRET = 'asdf'
+SECRET2 = 'ghjk'
+
+
+class AccountUpgradeTest(stubloader.StubbedTest):
+    def test_upgrade(self):
```
Comment (jonathanj) [inline axiom/test/historic/test_account2to3.py]: Should there be an assertion that `LoginAccount.password` is `None` now?

File: axiom/test/test_userbase.py [hunk 2]
```diff
@@ -637,52 +620,54 @@ def test_setPassword(self):
...
-            self._requestAvatarId(
-                UsernamePassword(username, self.password)),
-            UnauthorizedLogin)
+        d = account.setPassword(u'blahblah')
+        perform()
+        self.successResultOf(d)
+        d = self._requestAvatarId(UsernamePassword(username, u'blahblah'))
+        perform()
+        self.assertEquals(self.successResultOf(d), account.storeID)
+        d = self._requestAvatarId(UsernamePassword(username, self.password))
```
Comment (jonathanj) [inline axiom/test/test_userbase.py]: Might be worth adding another assertion that trying to authenticate with `account.passwordHash` will fail.

File: axiom/userbase.py [hunk 3]
```diff
@@ -283,8 +334,12 @@ def setPassword(self, newPassword):
 
         @return: A deferred firing when the password has been changed.
         """
-        self.password = None if newPassword is None else unicode(newPassword)
-        return succeed(None)
+        realm = self.store.findUnique(LoginSystem)
```
Comment (jonathanj) [inline axiom/userbase.py]: If `setPassword` or `replacePassword` is called on the user-store `LoginAccount` item, this `findUnique` query will fail for obvious reasons.

```
>>> from twisted.cred.portal import IRealm
>>> from axiom.userbase import LoginSystem, LoginAccount
>>> IRealm(db).accountByAddress(u'jonathan', u'localhost').avatars.open().findUnique(LoginAccount)
LoginAccount(avatars=reference(-1), disabled=0, password=None, passwordHash=u'$argon2i$v=19$m=102400,t=2,p=8$jpHSmnMOoXSOsVaqtRaiVA$g/nVGDUMYCdSpoLmL27rcw', storeID=1)@0x110d25610
>>> _.store.findUnique(LoginSystem)
Traceback (most recent call last):
  File "<axiom browser>", line 1, in <module>
  File "/Users/jonathan/.virtualenvs/axiom/lib/python2.7/site-packages/axiom/store.py", line 1631, in findUnique
    raise errors.ItemNotFound(comparison)
ItemNotFound: None
```

File: axiom/userbase.py [hunk 4]
```diff
@@ -31,19 +31,26 @@
...
+from threading import Thread
 
 from zope.interface import implements, Interface
 
 from twisted.cred.portal import IRealm
-from twisted.cred.credentials import IUsernamePassword, IUsernameHashedPassword
+from twisted.cred.credentials import IUsernamePassword
 from twisted.cred.checkers import ICredentialsChecker, ANONYMOUS
 from twisted.python import log
 from twisted.internet.defer import succeed, fail
```
Comment (jonathanj) [inline axiom/userbase.py]: I think nothing uses `succeed` now.

**Comment by glyph** [comment]
Hi @jsza !

I'm inclined to land this.  Is there any compelling reason not to? :)

**Comment by glyph** [comment]
specifically:

> should not be merged before the new APIs have been included in a release

a bit curious about what this means? which new APIs?

**Comment by mithrandi** [comment]
This refers to #91.

**Pull request merged**